### ISSUE TITLE
Editable ContentContainer Default Permissions

### ISSUE THREAD

**Issue by luke-** [humhub/humhub#4670]
Make ContentContainer (e.g. Space / User) Permissions editable by administators.

Space Admin:
![image](https://user-images.githubusercontent.com/4736168/100457738-e218d900-30c2-11eb-87d6-b11727b73a24.png)

User Admin:
![image](https://user-images.githubusercontent.com/4736168/100457903-27d5a180-30c3-11eb-829e-857fa2ecda1c.png)

### Is your feature request related to a problem? Please describe.

Configuration of default Space permissions is rather complex.
e.g. https://docs.humhub.org/docs/admin/permissions

### Describe the solution you'd like

Permission editor grid on Space settings menu

**Comment by yurabakhtin** [comment]
@luke- It seems you attached the same image for "User Admin:" as for "Space Admin, please confirm you meant this:
<img width="1153" alt="user_admin" src="https://user-images.githubusercontent.com/6995524/100569828-45338700-32e0-11eb-8e5f-05c736b869df.png">
"

**Comment by luke-** [comment]
@yurabakhtin Oh, yes! This would of course be the best place for the profile default settings.

**Comment by yurabakhtin** [comment]
@luke- https://github.com/humhub/humhub/pull/4678/commits/62f14bb3cc70a0a4c5d8512a0ee3af1c30748875 Please see how I implemented default permissions for Spaces:
<img width="1148" alt="prview_space_default_perms" src="https://user-images.githubusercontent.com/6995524/100609830-b9d9e600-331f-11eb-8870-ec94d5b9a803.png">

As you can see many permissions are disabled. I specially did this because I think almost all modules has a not completed initialization for permissions.
Only single core module was implemented enough, please see here what I mean https://github.com/humhub/humhub/blob/enh/4670-editable_contentcontainer_default_permissions/protected/humhub/modules/post/Module.php#L32.
`new permissions\CreatePost(['contentContainer' => $contentContainer])` - the Content Container is passed there, so we can know what Model(Space or User) where have there.
Also you can see on my screenshot a Space permission is active only, it is why I did this here https://github.com/humhub/humhub/commit/62f14bb3cc70a0a4c5d8512a0ee3af1c30748875#diff-2672f0ad55d9b222ba975092e5439006c0058e175b02c4a40cecc70eef4d601bR70.
But all other modules don't initialize the Content Container, I suggest to modify ALL modules, firstly only core and then all external modules to do this fix. After that we can edit all default permissions.

---
Please review new DB table for default permissions:
<img width="1027" alt="default_perm_table" src="https://user-images.githubusercontent.com/6995524/100610525-f2c68a80-3320-11eb-8ca0-b427d1a2d02d.png">
There is main column is `contentcontainer_class` in order to know what permission type we have there, either `\Space` or `\User`, I think it is what was required, I mean to keep different default permissions for Space and for User.
But please note config from file has no such separate thing. As I understand in order to don't mix permission in config file the models `Space` and `User` has unique keys, e.g. `User` has a prefix `u_` for `User::USERGROUP_USER == 'u_user'`, but `Space` has no prefixes so `Space::USERGROUP_USER == 'user'`.

I am not sure that is good, but if it some rule to keep the keys unique then probably we can remove the column `contentcontainer_class` and keep PK only as `permission_id, group_id, module_id` because `group_id` is(may be) unique between `Space` and `User`, and if we will do this then we can keep the permissions initialization in Modules as is(without passing `['contentContainer' => $contentContainer]`), but I don't like this solution. Please tell me what you think, thanks.

---
Not important:
Check my this code https://github.com/humhub/humhub/commit/62f14bb3cc70a0a4c5d8512a0ee3af1c30748875#diff-6237739394f8e07525b453a79b6e23499f3b1184d0461ac8735a1f2e10df3733R149: `$groups = (new Space())->getUserGroups();`. I think it would be better to change the method to static because it looks like a config array. But I am afraid to touch such core method because it may be used in external module. Only if you confirm. Another way is to move the array to some new static method of the Space Module or keep as I did if you think it is good code.

**Comment by luke-** [comment]
@yurabakhtin 

initialization: I also think your solution is nicer, but then we should adapt all existing permissions and if necessary also update the documentation to make sure that the permissions are initialized correctly.

GetUserGroups: We can mark them as static. 

Caching: Since the default state probably has to be checked often, we should pay special attention to caching and try to minimize DB queries.

Wording: Here I found the wording a littl bit misleading. The defaults do not only affect new spaces. Maybe a warning that this also affects existing spaces would be good. But the behavior is good and wanted.  thttps://github.com/humhub/humhub/commit/62f14bb3cc70a0a4c5d8512a0ee3af1c30748875#diff-6237739394f8e07525b453a79b6e23499f3b1184d0461ac8735a1f2e10df3733R149

**Comment by yurabakhtin** [comment]
@luke- Default permissions for User Containers - https://github.com/humhub/humhub/commit/dee104293d892a46b7514d93ed4f71914130385d:
<img width="1148" alt="default_user_permissions" src="https://user-images.githubusercontent.com/6995524/100723111-6ffe0800-33d2-11eb-8690-db1e90922ac6.png">

**Comment by luke-** [comment]
Looks fine!

As next you fixing the "disabled" permissions, right? 
What also be good would be some unit tests for it.

**Comment by yurabakhtin** [comment]
@luke- 

> As next you fixing the "disabled" permissions, right?

Yes, I did this here https://github.com/humhub/humhub/pull/4678/commits/79695b56ee11243db6d88ac6a145e1778d12e70f, but only for core/built-in modules:
<img width="854" alt="core_user_default_perms" src="https://user-images.githubusercontent.com/6995524/100728540-bd7d7380-33d8-11eb-9535-0ad59bc89e4e.png">
<img width="854" alt="core_space_default_perms" src="https://user-images.githubusercontent.com/6995524/100728810-0f25fe00-33d9-11eb-925e-4f17d366512d.png">

Next point is to do same at least for most popular external modules, could you please tell me what external modules should be updated in first order?

> What also be good would be some unit tests for it.

Ok, I will see what we can do.

**Comment by yurabakhtin** [comment]
> GetUserGroups: We can mark them as static.

Commit https://github.com/humhub/humhub/pull/4678/commits/afec885c8a7aa8455b4d9c5fd0ee24bf02bddd3a

**Comment by yurabakhtin** [comment]
> Wording: Here I found the wording a littl bit misleading. The defaults do not only affect new spaces. Maybe a warning that this also affects existing spaces would be good. But the behavior is good and wanted.

Ah yes, you are right, I removed the word "new" from description "Here you can define default permissions for new spaces." - https://github.com/humhub/humhub/commit/10c0a9217119a33f52ce11c2d478d93bd3a30bb1.

**Comment by yurabakhtin** [comment]
@luke- 

> What also be good would be some unit tests for it.

Commit https://github.com/humhub/humhub/pull/4678/commits/7546cd8593a3d2324d2fa5f2a2eecb40ce4c99cb

**Comment by luke-** [comment]
@yurabakhtin Is it problematic for backward compatibility to start fixes/pull requests for the modules now? Then all modules are ready when 1.8 will be released in March.

**Comment by yurabakhtin** [comment]
@luke- I don't see any problem to start to the fixes like I did here https://github.com/humhub/humhub/pull/4678/commits/79695b56ee11243db6d88ac6a145e1778d12e70f for all other external modules. I just asked you to get a priority list of modules to know what should be updated firstly, but if it doesn't matter then I can start it from any first found module.

**Comment by luke-** [comment]
@yurabakhtin There is no priority list, otherwise humhub/* and then humhub-contrib/*

**Comment by yurabakhtin** [comment]
@luke-

> Caching: Since the default state probably has to be checked often, we should pay special attention to caching and try to minimize DB queries.

I have implemented this in commit https://github.com/humhub/humhub/commit/55b008da00096d6686b9ae690b6883ffcc2d8327.
I decided to split and store in 2 different vars:

- `defaultPermissions:humhub\modules\space\models\Space `
- `defaultPermissions:humhub\modules\user\models\User`

Each of them is an array with structure `[group_id][module_id][permission_id]`.
Of course we can keep all in single array, but I think it is ok to have 2 separate queries, because if I am right it is very rarely to keep on the same page permissions and for Space and for User Containers.

Please note I did to clear a cache array(Space or User) completely on updating of each default permission.

**Comment by yurabakhtin** [comment]
@luke- Please see on the screenshot below and let me know if I should investigate where the column is used, so probably it can be deleted completely:
<img width="1137" alt="same_data_column" src="https://user-images.githubusercontent.com/6995524/100833529-333a1b80-347b-11eb-8f9d-653ba96d1ea1.png">

**Comment by yurabakhtin** [comment]
@luke- 

> Is it problematic for backward compatibility to start fixes/pull requests for the modules now? Then all modules are ready when 1.8 will be released in March.
> There is no priority list, otherwise humhub/* and then humhub-contrib/*

I have checked each module and did the changes where a module has defined permissions, you can find full list in the PR https://github.com/humhub/humhub/pull/4678#issue-529510086

**Comment by buddh4** [comment]
@luke- @yurabakhtin as far is I know currently its even allowed to only return a string class in the `Module::getPermission()`, I'am not sure why we need to set the conentcontainer here could you explain why we needed to change this? If a string class (or permission without contencontainer) is not allowed anymore this is should definetly be mentioned in the migration guide etc and maybe discussed and tested.

**Comment by luke-** [comment]
@yurabakhtin The differentiation between ID and class is necessary for permissions, of which there are several variants.   An example would be the space types in Enterprise Theme. Here you can define if a user is allowed to create spaces per type.
https://git.humhub.org/professional/enterprise-theme/-/blob/master/Module.php#L78-92

**Comment by buddh4** [comment]
I've thought about this some more and here are my concerns:

- `BasePermission::$contentContainer` is currently only used for description/title generation of permission
- `BasePermission::$contentContainer` is currently only used in one place `humhub\modules\post\permissions\CreatePost`

This field was introduced as workaround to avoid a heavy breaking change (to add $container property to `getDescription()` and `getTitle()`) but
its not really a solid design. We should not push people forward to use this (if not needed for description/title) and do not force
modules to use this in `Module::getPermissions()`. Permission instances should stay stateless and not directly bound to a container instance especially.

Usually permissions are verified by something like this: `$spaceXY->can(CreatePost::class)` or `$spaceXY->can(new CreatePost)` this is widely used and
here we do not use a container relation neither. The permission manager is responsible for connecting the user / permission / container and the permission.

Conclusion:

- The property was rather a bad design choice (workaround) and should not be pushed forward
- We should not force or encourage people to use it unless they need to for tile / description
- Permission instances should stay stateless and just provide metadata for the permission manager which does the actual work

**Comment by luke-** [comment]
@buddh4 Good points! Marking the existing attribute as deprecated and switching to an interface would probably a good solution.

Anyway, I would suggest we create a new issue for this and merge this PR (ofc with reference in the new issue). 
 The PR does not add new APIs or other breaking changes. Only for permissions that are specified without a ContentContainer, it is not possible to set defaults atm.

**Comment by buddh4** [comment]
@luke- Yes, the `ContainerPermission` interface can be done in another PR, but moving the default value loading out of `BasePermission` and cut the dependency to `BasePermission::$contentContainer` needs to be done prior to migrating it. It is an API change when now `BasePermission::$contentContainer` is involved in the actual permission verification process (which it wasn't prior to this change).

Imagine:

```
$permission = new MyPermission(['contentContainer' => $xy]);

$space1->can($permission);
$space2->can($permission);

// Whats the result?
$permission = $permission->contentContainer;
```

**Comment by yurabakhtin** [comment]
@luke- 

> The differentiation between ID and class is necessary for permissions, of which there are several variants. An example would be the space types in Enterprise Theme. Here you can define if a user is allowed to create spaces per type.
https://git.humhub.org/professional/enterprise-theme/-/blob/master/Module.php#L78-92

Ok, If you think that is normal and the column `class` is really need for different purpose.

**Comment by yurabakhtin** [comment]
@buddh4 I have read your comment above, thank you for explanations. Yes, I agree it was not good solution to use `BasePermission::$contentContainer` only in order to make default permission editable.
As I understand we should forgot to use the property `BasePermission::$contentContainer` and also move `BasePermission->getDefaultStoredState($groupId)` out of `BasePermission` probably into `ContentContainerDefaultPermissionManager`, right?
Should I fix this inside PR https://github.com/humhub/humhub/pull/4678 or it will done in some other new PR?

**Comment by luke-** [comment]
@yurabakhtin I would prefer to implement and merge this default permissions PR first. 
Then we can start a new PR, improve the ContentContainerPermissions logc and also improve the parts of this PR.  
But we don't have to do unnecessary work for that either. So it's your call.

**Comment by buddh4** [comment]
@luke- Why do you prefer to merging a PR which is not finished and will not implement the solution which will be part of 1.8?

**Comment by luke-** [comment]
@buddh4 Thought the PR is almost finished and working. 
In case I have misunderstood, then into this PR or another one in advance.

**Comment by buddh4** [comment]
kk, thought you want to merge it in this state and refactor in another PR for some reason. We only have to move the default logic out of BasePermission and move it ti ContentcontainerPermissionManager.

**Comment by yurabakhtin** [comment]
@luke- I agree with @buddh4, the PR doesn't work correctly as I described here https://github.com/humhub/humhub/pull/4678#issuecomment-738595064 after my retesting. Indeed this is my bad of not enough testing.

**Comment by yurabakhtin** [comment]
@luke- I have implemented a solution advised by @buddh4, more details here https://github.com/humhub/humhub/pull/4678#issuecomment-738632155. Only after that I think this PR can be merged, because before it didn't work properly.
Many thanks for @buddh4 👍

**Comment by luke-** [comment]
@yurabakhtin All right! So if I understood it right, we do this all in PR #4768.

**Comment by yurabakhtin** [comment]
@luke- Yes right, the fix/solution was done in the same PR https://github.com/humhub/humhub/pull/4678 as I mentioned in my comment above.

**Comment by yurabakhtin** [comment]
@luke- This issue can be closed.
